File: LeetCode/34FindFirstAndLastPositionOfElementInSortedArray/FindFirstAndLastPositionOfElementInSortedArray.cpp
class Solution {
public:
    int binarySearch(vector<int>& nums, int target, bool check_left) {
        int start = 0, end = nums.size()-1, i = -1, mid;
        while (start <= end) {
            mid = (start + end) / 2;
            if (target > nums[mid]) {
                start = mid + 1;
            }
            else if (target < nums[mid]) {
                end = mid - 1;
            }
            else {
                i = mid;
                if (check_left) {
                    end = mid - 1;
                }
                else {
                    start = mid + 1;
                }
            }
        }
        return i;
    }
    vector<int> searchRange(vector<int>& nums, int target) {
        // Binary Search -> Time:O(log(n)), Space:O(1)
        vector<int> ans;
        ans.push_back(binarySearch(nums, target, true));
        ans.push_back(binarySearch(nums, target, false));
        return ans;
    }
};

File: LeetCode/34FindFirstAndLastPositionOfElementInSortedArray/FindFirstAndLastPositionOfElementInSortedArray.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        # Binary Search -> Time:O(logn), Space:O(1)
        def binary_search(nums, target, check_left=True):
            start, end = 0, len(nums)-1
            i = -1
            while start <= end:
                mid = (start+end)//2
                if target > nums[mid]:
                    start = mid + 1
                elif target < nums[mid]:
                    end = mid - 1
                else:
                    i = mid
                    if check_left:
                        end = mid - 1
                    else:
                        start = mid + 1
            return i
        return [binary_search(nums, target), binary_search(nums, target, False)]
